### PR TITLE
Feature/custom operations

### LINKED ISSUE #279: Custom operations
Epic: https://github.com/ZupIT/beagle/issues/884

Implement a way to specify custom operations in the BeagleConfig. Please, read the epic for more specific details on this feature.

### PR DESCRIPTION
**Pull request by hectorcustodiozup**
closes #279 
**- What I did**
Added new option 'customOperations' to BeagleConfig interface
**- How I did it**
Created new property customOperation
Validated operations names and cast warnings when needed
Concatenated the default list of operations with the one given through the config

**- How to verify it**
Create custom operations and use them in expressions
**- Description for the changelog**
New option 'customOperations'

### CODE REVIEW

File: src/service/beagle-service/types.ts [hunk 1]
```diff
@@ -130,6 +132,10 @@ export interface BeagleConfig<Schema> {
    * options, use `default: true`.
    */
   navigationControllers?: Record<string, NavigationController>,
+  /**
```
Comment (Tiagoperes) [inline src/service/beagle-service/types.ts:135]: This is the documentation that appears to the user when he's programming. Please give more details:

Custom operations can be used to extend the capability of the Beagle expressions and are called like functions, e.g. `@{sum(1, 2)}`. Beagle has already many default operations, this configuration can be used to extend this initial set. The keys of this object represent the operation name and the values must be the functions themselves. An operation name must contain only letters, numbers and the character `_`, it also must contain at least one letter or `_`.

PS: You can write github markdown in these documentations, so, you can use the code delimiter (`) freely.

File: src/beagle-view/render/expression.ts [hunk 2]
```diff
@@ -91,21 +93,22 @@ function parseParameters(parameterString: string) {
...
+  if (!operationHandlers[operationName]) {
     throw new BeagleNotFoundError(`operation with name "${operationName}" doesn't exist.`)
   }
   const params = parseParameters(paramString)
   // eslint-disable-next-line
-  const resolvedParams: any[] = params.map(param => evaluateExpression(param, contextHierarchy))
+  const resolvedParams: any[] = params.map(param => evaluateExpression(param, contextHierarchy, operationHandlers))
 
-  const fn = Operation[operationName] as (...args: any) => any
+  const fn = operationHandlers[operationName] as (...args: any) => any
```
Comment (Tiagoperes) [inline src/beagle-view/render/expression.ts:111]: use the type `Operation` here

**Review by Tiagoperes — CHANGES_REQUESTED** [review]
I didn't check the tests yet. I'll only review the tests when the implementation is ok.

File: src/service/beagle-service/configuration.ts [hunk 3]
```diff
@@ -62,15 +62,15 @@ function getLifecycleHookMap(
...
 
-    const match = key.match(/^[A-Za-z]{1}$|^[a-zA-Z0-9_]*[A-Za-z_]{1,}[0-9_]*$/)
-    if (!match) {
-      delete operations[key]
-      logger.warn(`Please, make sure your operation names contain only letters, numbers and the symbol "_" , 
-      any operation not following this rule will be ignored "${key}"`)
-    }
+    const match = key.match(/^\w*[A-z_]+\w*$/)
+    if (!match)
+      throw new BeagleError(`Operation names must have only letters, numbers and the character _.
```
Comment (Tiagoperes) [inline src/service/beagle-service/configuration.ts:70]: Infelizmente, quebras de linha e tabs aqui resultam em quebras de linhas e tabs na string final. Como não queremos replicar a formatação do nosso código na string de erro, não podemos colocar nem quebra de linha nem tabs aqui.

Por favor, coloque tudo na mesma linha.

File: src/service/beagle-service/types.ts [hunk 4]
```diff
@@ -133,7 +133,12 @@ export interface BeagleConfig<Schema> {
    */
   navigationControllers?: Record<string, NavigationController>,
   /**
-   * The map of custom operations.
+   * The map of custom operations that can be used to extend the capability of the Beagle expressions and are called like functions, 
+   * e.g. @{sum(1, 2)}.
+   * The keys of this object represent the operation name and the values must be the functions themselves. 
+   * An operation name must contain only letters, numbers and the character _, 
+   * it also must contain at least one letter or _.
+   * Note: If you create custom operations using the same name of a default from Beagle, the default will be overwritten by the custom one
```
Comment (Tiagoperes) [inline src/service/beagle-service/types.ts:141]: use ` para delimitar código. Por exemplo:
```
`@{sum(1, 2)}`
```

**Review by Tiagoperes — CHANGES_REQUESTED** [review]

File: src/service/beagle-service/configuration.ts (hunk 3 above)
Comment (hectorcustodiozup) [reply src/service/beagle-service/configuration.ts]: Nossa, não sabia disso. achei que o lint reclamaria pelo tamanho da linha.

File: src/service/beagle-service/types.ts (hunk 4 above)
Comment (hectorcustodiozup) [reply src/service/beagle-service/types.ts:141]: Coloquei só no que você sugeriu, não vi outra necessidade

**Review by Tiagoperes — APPROVED** [review]

**Review by carolinegoncalveszup — APPROVED** [review]

**Pull request merged**